server/fleet/datastore.go

```diff
@@ -398,16 +398,17 @@ type Datastore interface {
 	ListSoftwareVulnerabilitiesByHostIDsSource(ctx context.Context, hostIDs []uint, source VulnerabilitySource) (map[uint][]SoftwareVulnerability, error)
 	LoadHostSoftware(ctx context.Context, host *Host, includeCVEScores bool) error
 
-	// ListSoftwareBySourceIter returns an iterator for consuming all software rows filtered by
-	// their 'source'.
-	ListSoftwareBySourceIter(ctx context.Context, sources []string) (SoftwareIterator, error)
-
-	AllSoftwareWithoutCPEIterator(ctx context.Context, excludedPlatforms []string) (SoftwareIterator, error)
-	AddCPEForSoftware(ctx context.Context, software Software, cpe string) error
+	AllSoftwareIterator(ctx context.Context, query SoftwareIterQueryOptions) (SoftwareIterator, error)
+	// UpsertSoftwareCPEs either inserts new 'software_cpe' entries, or if a now with the same CPE
+	// already exists, performs an update operation. Returns the number of rows affected.
+	UpsertSoftwareCPEs(ctx context.Context, cpes []SoftwareCPE) (int64, error)
+	// DeleteSoftwareCPEs removes entries from 'software_cpe' by matching the software_id in the
+	// provided cpes. Returns the number of rows affected.
+	DeleteSoftwareCPEs(ctx context.Context, cpes []SoftwareCPE) (int64, error)
 	ListSoftwareCPEs(ctx context.Context) ([]SoftwareCPE, error)
 	// InsertSoftwareVulnerabilities inserts the given vulnerabilities in the datastore, returns the number
```

Comment:
The doc string needs update. (Guessing now it returns true if there was an insert, not an upsert.)

Reply:
Done in https://github.com/fleetdm/fleet/commit/168c91150ef4691225d699b4a2f5b9ff6abf9245